File: goodrouter-net/Goodrouter/Router.cs
using System.Text.RegularExpressions;

/// <summary>
/// This is the router!
/// </summary>
public class Router<K> where K : notnull
{
    private RouteNode<K> rootNode = new RouteNode<K>();
    private readonly Dictionary<K, RouteNode<K>> leafNodes = new Dictionary<K, RouteNode<K>>();

    private Regex parameterPlaceholderRE = new Regex("\\{(.*?)\\}");
    private int maximumParameterValueLength = 20;

    private Func<string, string> parameterValueEncoder =
        (string decodedValue) => Uri.EscapeDataString(decodedValue);

    private Func<string, string> parameterValueDecoder =
        (string encodedValue) => Uri.UnescapeDataString(encodedValue);

    /// <summary>
    /// Set the maximum length of an encoded parameter value
    /// </summary>
    /// <param name="value">
    /// The maximum length
    /// </param>
    /// <returns>
    /// Router object, so you can chain!
    /// </returns>
    public Router<K> SetMaximumParameterValueLength(int value)
    {
        maximumParameterValueLength = value;
        return this;
    }

    /// <summary>
    /// Set the regular expression that will be used to parse placeholders from a route template
    /// </summary>
    /// <param name="value">
    /// The regular expression
    /// </param>
    /// <returns>
    /// Router object, so you can chain!
    /// </returns>
    public Router<K> SetParameterPlaceholderRE(Regex value)
    {
        parameterPlaceholderRE = value;
        return this;
    }

    /// <summary>
    /// Sets the function that will be used to encode parameter values
    /// </summary>
    /// <param name="value">
    /// Function to use
    /// </param>
    /// <returns>
    /// Router object, so you can chain!
    /// </returns>
    public Router<K> SetParameterValueEncoder(Func<string, string> value)
    {
        parameterValueEncoder = value;
        return this;
    }

    /// <summary>
    /// Sets the function that will be used when decoding parameter values
    /// </summary>
    /// <param name="value">
    /// Function to use
    /// </param>
    /// <returns>
    /// Router object, so you can chain!
    /// </returns>
    public Router<K> SetParameterValueDecoder(Func<string, string> value)
    {
        parameterValueDecoder = value;
        return this;
    }

    /// <summary>
    /// Adds a new route
    /// </summary>
    /// <param name="routeKey">
    /// name of the route
    /// </param>
    /// <param name="routeTemplate">
    /// template for the route, als defines parameters
    /// </param>
    public Router<K> InsertRoute(
        K routeKey,
        string routeTemplate
    )
    {
        var leafNode = this.rootNode.Insert(
            routeKey,
            routeTemplate,
            this.parameterPlaceholderRE
        );
        this.leafNodes.Add(routeKey, leafNode);
        return this;
    }

    /// <summary>
    /// Match the path against one of the provided routes and parse the parameters in it
    /// </summary>
    /// <param name="path">
    /// path to match
    /// </param>
    /// <returns>
    /// route that is matches to the path or null if no match is found
    /// </returns>
    public (K?, IReadOnlyDictionary<string, string>) ParseRoute(
        string path
    )
    {
        var parameters = new Dictionary<string, string>();

        var (routeKey, parameterNames, parameterValues) = this.rootNode.Parse(
            path,
            this.maximumParameterValueLength
        );

        for (var index = 0; index < parameterNames.Count; index++)
        {
            var parameterName = parameterNames[index];
            var parameterValue = parameterValues[index];
            parameters[parameterName] = parameterValueDecoder(parameterValue);
        }

        return (
            routeKey,
            parameters
        );
    }

    /// <summary>
    /// Convert a route to a path string.
    /// </summary>
    /// <param name="routeKey">
    /// route to stringify
    /// </param>
    /// <returns>
    /// string representing the route or null if the route is not found by name
    /// </returns>
    public string? StringifyRoute(
        K routeKey
    )
    {
        return StringifyRoute(routeKey, new Dictionary<string, string>());
    }
    /// <summary>
    /// Convert a route to a path string.
    /// </summary>
    /// <param name="routeKey">
    /// route to stringify
    /// </param>
    /// <param name="routeParameters">
    /// parameters for the route
    /// </param>
    /// <returns>
    /// string representing the route or null if the route is not found by name
    /// </returns>
    public string? StringifyRoute(
        K routeKey,
        IReadOnlyDictionary<string, string> routeParameters
    )
    {
        var node = this.leafNodes[routeKey];
        if (node == null) return null;

        var parameterValues = new List<string>();
        foreach (var parameterName in node.RouteParameterNames)
        {
            var parameterValue = routeParameters[parameterName];
            parameterValues.Add(parameterValueEncoder(parameterValue));
        }

        return node.Stringify(parameterValues);
    }

}

File: goodrouter-ts/src/router.ts
import { RouteNodeJson } from "./route-node-json.js";
import { RouteNode } from "./route-node.js";
import { defaultRouterOptions, RouterOptions } from "./router-options.js";
import { parseTemplatePairs } from "./template.js";

/**
 * @description
 * This is the actual router that contains all routes and does the actual routing
 *
 * @example
 * ```typescript
 * const router = new Router<string>();
 *
 * router.insertRoute("all-products", "/product/all");
 * router.insertRoute("product-detail", "/product/{id}");
 *
 * // And now we can parse routes!
 *
 * {
 *   const [routeKey, routeParameters] = router.parseRoute("/not-found");
 *   assert.equal(routeKey, null);
 *   assert.deepEqual(routeParameters, {});
 * }
 *
 * {
 *   const [routeKey, routeParameters] = router.parseRoute("/product/all");
 *   assert.equal(routeKey, "all-products");
 *   assert.deepEqual(routeParameters, {});
 * }
 *
 * {
 *   const [routeKey, routeParameters] = router.parseRoute("/product/1");
 *   assert.equal(routeKey, "product-detail");
 *   assert.deepEqual(routeParameters, { id: "1" });
 * }
 *
 * // And we can stringify routes
 *
 * {
 *   const path = router.stringifyRoute(
 *     "all-products",
 *   });
 *   assert.equal(path, "/product/all");
 * }
 *
 * {
 *   const path = router.stringifyRoute(
 *     "product-detail",
 *     { id: "2" },
 *   );
 *   assert.equal(path, "/product/2");
 * }
 * ```
 */
export class Router<K extends string | number> {
    constructor(options: RouterOptions = {}) {
        this.options = {
            ...defaultRouterOptions,
            ...options,
        };
    }

    protected options: RouterOptions & typeof defaultRouterOptions;

    private rootNode = new RouteNode<K>();
    private leafNodes = new Map<K, RouteNode<K>>();
    private templatePairs = new Map<
        K,
        Array<readonly [string, string | null]>
    >();

    /**
     * @description
     * Adds a new route
     *
     * @param routeKey name of the route
     * @param routeTemplate template for the route, als defines parameters
     */
    public insertRoute(routeKey: K, routeTemplate: string) {
        const templatePairs = [
            ...parseTemplatePairs(
                routeTemplate,
                this.options.parameterPlaceholderRE,
            ),
        ];
        this.templatePairs.set(routeKey, templatePairs);
        const leafNode = this.rootNode.insert(routeKey, templatePairs);
        this.leafNodes.set(routeKey, leafNode);
        return this;
    }

    /**
     * @description
     * Match the path against a known routes and parse the parameters in it
     *
     * @param path path to match
     * @returns tuple with the route name or null if no route found. Then the parameters
     */
    public parseRoute(path: string): [K | null, Record<string, string>] {
        const parameters: Record<string, string> = {};

        const [routeKey, parameterValues] = this.rootNode.parse(
            path,
            this.options.maximumParameterValueLength,
        );

        if (routeKey == null) {
            return [null, {}];
        }

        const templatePairs = this.templatePairs.get(routeKey);
        if (templatePairs == null) {
            // this never happens
            return [null, {}];
        }

        for (let index = 0; index < parameterValues.length; index++) {
            const [, parameterName] = templatePairs[index + 1];
            if (parameterName == null) {
                // this never happens
                return [null, {}];
            }
            const parameterValue = parameterValues[index];
            parameters[parameterName] =
                this.options.parameterValueDecoder(parameterValue);
        }

        return [routeKey, parameters];
    }

    /**
     * @description
     * Convert a route to a path string.
     *
     * @param routeKey route to stringify
     * @param routeParameters parameters to include in the path
     * @returns string representing the route or null if the route is not found by name
     */
    public stringifyRoute(
        routeKey: K,
        routeParameters: Record<string, string> = {},
    ): string | null {
        let result = "";
        const templatePairs = this.templatePairs.get(routeKey);
        if (templatePairs == null) {
            return null;
        }
        for (let index = 0; index < templatePairs.length; index++) {
            const [parameterAnchor, parameterName] = templatePairs[index];
            if (parameterName != null) {
                const parameterValue = routeParameters[parameterName];
                result += this.options.parameterValueEncoder(parameterValue);
            }
            result += parameterAnchor;
        }
        return result;
    }

    public saveToJSON(): RouteNodeJson<K> {
        const json = this.rootNode.toJSON();
        return json;
    }

    public loadFromJSON(json: RouteNodeJson<K>) {
        const node = RouteNode.fromJSON(json);
        this.rootNode = node;
        return this;
    }
}
